File: United_Community/Mradul/Daily-Leetcode/day_16_CountAndSay.java
// https://leetcode.com/problems/count-and-say/


class Solution {
    public String countAndSay(int n) {
        if (n ==  1) {
            return "1"; // condition for recursion
        }
        return (say(countAndSay(n-1))); // calling function inside a function
    }

    public static String say (String say) {
        int len = say.length(), count;
        StringBuilder sb = new StringBuilder("");

        for (int i = 0; i < len; i++) {
            count = 1;
            while (i < len - 1 && say.charAt(i) == say.charAt(i+1)) { //geting the count
                count++;
                i++;
            }
            sb.append(count).append(say.charAt(i)); // adding the count and the cahr
        }
        return sb.toString(); // out put what will thr string will be said as
    }
}

File: United_Community/Amit/Daily-Leetcode/day_16_count_and_say.cpp
class Solution {
public:
    string countAndSay(int n) {
        vector<string> result;
        string s = "";
        result.push_back(s);

        for (int i = 1; i <= n; i++) {
            s = result[i - 1];
            if (s == "") {
                result.push_back("1");
            } else {
                string new_res = "";
                int count = 1;

                for (int j = 0; j < s.size() - 1; j++) {
                    if (s[j + 1] == s[j]) {
                        count++;
                    } else {
                        new_res += (count + '0');
                        new_res += s[j];
                        count = 1;
                    }
                }

                new_res += (count + '0');
                new_res += s[s.size() - 1];
                result.push_back(new_res);
            }
        }
        return result[n];
    }
};
